catalog/src/main/java/org/killbill/billing/catalog/VersionedCatalog.java

```diff
@@ -182,21 +184,29 @@ private CatalogPlanEntry findCatalogPlanEntry(final PlanRequestWrapper wrapper,
             }
 
 
-            final boolean initialVersion = (i == 0);
+            final boolean oldestCatalog = (i == 0);
             final DateTime catalogEffectiveDate = CatalogDateHelper.toUTCDateTime(c.getEffectiveDate());
-            if (initialVersion || // Prevent issue with time granularity -- see #760
-                !subscriptionStartDate.isBefore(catalogEffectiveDate)) { // It's a new subscription this plan always applies
+            final boolean catalogOlderThanSubscriptionStartDate = !subscriptionStartDate.isBefore(catalogEffectiveDate);
+            if (oldestCatalog || // Prevent issue with time granularity -- see #760
+                catalogOlderThanSubscriptionStartDate) { // It's a new subscription, this plan always applies
                 return new CatalogPlanEntry(c, plan);
-            } else { //Its an existing subscription
-                if (plan.getEffectiveDateForExistingSubscriptions() != null) { //if it is null any change to this does not apply to existing subscriptions
+            } else { // It's an existing subscription
+                if (plan.getEffectiveDateForExistingSubscriptions() != null) { // If it is null, any change to this catalog does not apply to existing subscriptions
                     final DateTime existingSubscriptionDate = CatalogDateHelper.toUTCDateTime(plan.getEffectiveDateForExistingSubscriptions());
-                    if (requestedDate.isAfter(existingSubscriptionDate)) { // this plan is now applicable to existing subs
+                    if (requestedDate.isAfter(existingSubscriptionDate)) { // This plan is now applicable to existing subs
                         return new CatalogPlanEntry(c, plan);
                     }
+                } else if (candidateInSubsequentCatalog == null) {
+                    // Keep the most recent one
+                    candidateInSubsequentCatalog = new CatalogPlanEntry(c, plan);
                 }
             }
         }
 
+        if (candidateInSubsequentCatalog != null) {
+            return candidateInSubsequentCatalog;
+        }
+
```

Comment:
What you have implemented should work as a fix but fundamentally this api is overloaded with respect to various use cases. At least those two categories:

- Billing operation where we need to find the right catalog entry (and therefore the right price): In this scenario, going through versions from most recent to oldest and taking into account `effectiveDateForExistingSubscriptions ` make sense
- For other subscription scenario -- let's say change Plan-- the logic is much simpler, we should just return the entry from the catalog version matching the requested date, i.e `catalogs.get(0). findPlan()` 

I am now wondering if the fix is not to call `public Plan createOrFindPlan(final PlanSpecifier spec,
                                 final PlanPhasePriceOverridesWithCallContext overrides,
                                 final DateTime requestedDate)` [here](https://github.com/killbill/killbill/blob/master/subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java#L411) instead of  `public Plan createOrFindPlan(final PlanSpecifier spec,
                                 final PlanPhasePriceOverridesWithCallContext overrides,
                                 final DateTime requestedDate,
                                 final DateTime subscriptionStartDate)` ?

Reply:
Agreed - I've made the change and verified that the APIs without the `subscriptionStartDate` are always called for subscription operations, except for internal billing operations (see also https://github.com/killbill/killbill/issues/784#issuecomment-390244107).

This is mainly true today, except for `rebuildTransitions`: https://github.com/killbill/killbill/blob/5d63b9b5e88f2c603046b9f60139a93caee7fbbd/subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBase.java#L767-L769 which is called in a lot of code paths (including `doChangePlan`). Because of this, I also need to keep my original fix, or the bug is still present: if `getEffectiveDateForExistingSubscriptions` returns `null` (not specified for a new introduced plan for instance), `findCatalogPlanEntry` never finds anything otherwise.

Intuitively anyways, regardless of whether the right API is used or not, it seems that the implementation of this method should not assume that `effectiveDateForExistingSubscriptions` is set. It would be interesting to write a test for instance with V0 (price P), V1 (price P1) and V2 (price P2) where `effectiveDateForExistingSubscriptions` is set only in V1, and see which plan is returned when computing billing events at different times.